library/std/src/sys/sgx/thread_parker.rs

```diff
@@ -0,0 +1,93 @@
+//! Thread parking based on SGX events.
+
+use super::abi::{thread, usercalls};
+use crate::io::ErrorKind;
+use crate::pin::Pin;
+use crate::ptr::{self, NonNull};
+use crate::sync::atomic::AtomicPtr;
+use crate::sync::atomic::Ordering::{Acquire, Relaxed, Release};
+use crate::time::Duration;
+use fortanix_sgx_abi::{EV_UNPARK, WAIT_INDEFINITE};
+
+const EMPTY: *mut u8 = ptr::invalid_mut(0);
+/// The TCS structure must be page-aligned, so this cannot be a valid pointer
+const NOTIFIED: *mut u8 = ptr::invalid_mut(1);
+
+pub struct Parker {
+    state: AtomicPtr<u8>,
+}
+
+impl Parker {
+    /// Construct the thread parker. The UNIX parker implementation
+    /// requires this to happen in-place.
+    pub unsafe fn new(parker: *mut Parker) {
+        unsafe { parker.write(Parker::new_internal()) }
+    }
+
+    pub(super) fn new_internal() -> Parker {
+        Parker { state: AtomicPtr::new(EMPTY) }
+    }
+
+    // This implementation doesn't require `unsafe` and `Pin`, but other implementations do.
+    pub unsafe fn park(self: Pin<&Self>) {
```

Comment:
Can you provide documentation of what properties the `park` and `unpark` functions are expected to provide?

Reply:
These functions are exposed publicly as [`thread::park`](https://doc.rust-lang.org/nightly/std/thread/fn.park.html) and [`Thread::unpark`](https://doc.rust-lang.org/nightly/std/thread/struct.Thread.html#method.unpark), respectively, where they are also documented.

Reply:
Sorry, I lost track of this PR. I meant security properties, but I guess having them in `Thread::park` and `Thread::unpark` suffices.